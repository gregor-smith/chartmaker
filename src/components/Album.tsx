import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
