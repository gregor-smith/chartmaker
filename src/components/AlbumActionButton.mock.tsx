const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
