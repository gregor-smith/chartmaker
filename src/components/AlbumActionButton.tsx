import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
