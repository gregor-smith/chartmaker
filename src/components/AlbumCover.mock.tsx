import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
