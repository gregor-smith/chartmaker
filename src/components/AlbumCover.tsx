import React, { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import { Album } from '@/types'


export type AlbumCoverProps = {
    album: Album
    size: string
    onDragStart: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

// the overlay is given an additional non-emotion class so it can be used as a
// selector in the container's style
const overlaySelector = 'overlay'

const baseOverlayStyle = cx(
    overlaySelector,
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const overlayStyle = cx(baseOverlayStyle, overlayClass)
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            [`:not(:hover) .${overlaySelector}`]: {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!album.placeholder) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!album.placeholder}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album.placeholder ? undefined : album.name}>
            {image}
            <div className={overlayStyle}>
                {children}
            </div>
        </div>
    )
}
