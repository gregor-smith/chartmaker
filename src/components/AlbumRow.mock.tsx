const mock: typeof import('@/components/AlbumRow') = {
    AlbumRow: ({ children }) =>
        <div className='mock-album-row'>
            {children}
        </div>
}


export const AlbumRow = mock.AlbumRow
