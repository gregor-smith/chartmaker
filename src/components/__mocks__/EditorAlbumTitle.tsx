import type { FC } from 'react'

import type { EditorAlbumTitleProps } from '@/components/EditorAlbumTitle'


export const EditorAlbumTitle: FC<EditorAlbumTitleProps> = ({ album, highlighted }) => {
    const json = JSON.stringify(album)
    return (
        <div className='mock-editor-album-title'>
            {`Album: ${json}`}
            {`Highlighted: ${highlighted}`}
        </div>
    )
}
