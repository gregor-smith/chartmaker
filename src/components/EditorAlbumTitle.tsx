import type { FC } from 'react'

import { AlbumTitle } from '@/components/AlbumTitle'
import type { DispatchProps } from '@/reducer'
import type { NamedAlbum } from '@/types'
import { RenameAlbumButton } from '@/components/RenameAlbumButton'
import { DeleteAlbumButton } from '@/components/DeleteAlbumButton'


export type EditorAlbumTitleProps = DispatchProps & {
    album: NamedAlbum
    highlighted: number | undefined
}


export const EditorAlbumTitle: FC<EditorAlbumTitleProps> = ({
    dispatch,
    album,
    highlighted
}) =>
    <AlbumTitle dispatch={dispatch}
            id={album.id}
            name={album.name}
            highlighted={highlighted === undefined
                ? undefined
                : album.id === highlighted}>
        <RenameAlbumButton dispatch={dispatch} id={album.id}/>
        <DeleteAlbumButton dispatch={dispatch} id={album.id}/>
    </AlbumTitle>
