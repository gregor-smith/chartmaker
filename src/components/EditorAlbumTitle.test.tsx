import { render } from 'react-dom'

import { EditorAlbumTitle } from '@/components/EditorAlbumTitle'
import type { NamedAlbum } from '@/types'

import { RenderContainer, ignore } from '@/test-utils/utils'


jest.mock('@/components/AlbumTitle')
jest.mock('@/components/RenameAlbumButton')
jest.mock('@/components/DeleteAlbumButton')


const container = new RenderContainer()


test.each([ undefined, 123, 456 ])('renders title and buttons', highlighted => {
    const album: NamedAlbum = {
        id: 123,
        name: 'Test album',
        url: 'http://test.com'
    }
    render(
        <EditorAlbumTitle dispatch={ignore} album={album} highlighted={highlighted}>
            Test children
        </EditorAlbumTitle>,
        container.element
    )
    expect(container.element).toMatchSnapshot()
})
