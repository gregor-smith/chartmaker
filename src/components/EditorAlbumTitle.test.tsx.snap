// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders title and buttons 1`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-album-title"
  >
    ID: 123
    Name: Test album
    Highlighted: undefined
    <div
      class="mock-rename-album-button"
    >
      ID: 123
    </div>
    <div
      class="mock-delete-album-button"
    >
      ID: 123
    </div>
  </div>
</div>
`;

exports[`renders title and buttons 2`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-album-title"
  >
    ID: 123
    Name: Test album
    Highlighted: true
    <div
      class="mock-rename-album-button"
    >
      ID: 123
    </div>
    <div
      class="mock-delete-album-button"
    >
      ID: 123
    </div>
  </div>
</div>
`;

exports[`renders title and buttons 3`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-album-title"
  >
    ID: 123
    Name: Test album
    Highlighted: false
    <div
      class="mock-rename-album-button"
    >
      ID: 123
    </div>
    <div
      class="mock-delete-album-button"
    >
      ID: 123
    </div>
  </div>
</div>
`;
